File: tools/glosh-remote-spike/app/src/main/java/com/glosh/remote/spike/guide/autopilot/AutopilotActionGate.java
package com.glosh.remote.spike.guide.autopilot;

import com.glosh.remote.spike.guide.accessibility.ScanGenerationGuard;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Confidence;
import com.glosh.remote.spike.guide.autopilot.AutopilotUiModel.MatchedTarget;
import com.glosh.remote.spike.guide.autopilot.AutopilotUiModel.TargetKey;

public final class AutopilotActionGate {
    public boolean authorize(
            ScanGenerationGuard.Token token,
            ScanGenerationGuard.Token currentToken,
            MatchedTarget target,
            TargetKey expectedKey) {
        return token != null
                && token.equals(currentToken)
                && target != null
                && target.key() == expectedKey
                && target.confidence() == Confidence.HIGH
                && target.clickable()
                && target.unique()
                && target.marginOk();
    }
}

File: docs/artifacts/remote-install-autopilot/reference/AutopilotActionGate.kt
package reference.autopilot

data class ActionToken(val windowId:Int,val generation:Long,val fingerprint:String)

class AutopilotActionGate {
    fun authorize(
        token: ActionToken,
        currentToken: ActionToken,
        target: MatchedTarget?,
        expectedKey: TargetKey
    ): Boolean {
        if (token != currentToken) return false
        if (target == null || target.key != expectedKey) return false
        if (target.confidence != Confidence.HIGH) return false
        if (!target.clickable || !target.unique || !target.marginOk) return false
        return true
    }
}
